test/e2e/framework/podlogs/podlogs.go

```diff
@@ -0,0 +1,172 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package podlogs enables live capturing of all events and log
+// messages for some or all pods in a namespace as they get generated.
+// This helps debugging both a running test (what is currently going
+// on?) and the output of a CI run (events appear in chronological
+// order and output that normally isn't available like the command
+// stdout messages are available).
+package podlogs
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"io"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// LogsForPod starts reading the logs for a certain pod. If the pod has more than one
+// container, opts.Container must be set. Reading stops when the context is done.
+func LogsForPod(ctx context.Context, cs clientset.Interface, ns, pod string, opts *v1.PodLogOptions) (io.ReadCloser, error) {
+	req := cs.Core().Pods(ns).GetLogs(pod, opts)
+	return req.Context(ctx).Stream()
+}
+
+// CopyAllLogs follows the logs of all containers in all pods and writes each log line
+// with the pod/container name as prefix. It does that until the context is done or
+// until an error occurs.
+func CopyAllLogs(ctx context.Context, cs clientset.Interface, ns string, to io.Writer) error {
+	watcher, err := cs.Core().Pods(ns).Watch(meta.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "cannot create Pod event watcher")
+	}
+
+	go func() {
+		var m sync.Mutex
+		logging := map[string]bool{}
+		check := func() {
+			m.Lock()
+			defer m.Unlock()
+
+			pods, err := cs.Core().Pods(ns).List(meta.ListOptions{})
+			if err != nil {
+				fmt.Fprintf(to, "get pod list in %s: %s", ns, err)
+				return
+			}
+
+			for _, pod := range pods.Items {
+				for _, c := range pod.Spec.Containers {
+					name := pod.ObjectMeta.Name + "/" + c.Name
+					if logging[name] {
+						continue
+					}
+					readCloser, err := LogsForPod(ctx, cs, ns, pod.ObjectMeta.Name,
+						&v1.PodLogOptions{
+							Container: c.Name,
+							Follow:    true,
+						})
+					if err != nil {
+						fmt.Fprintf(to, "%s: %s\n", name, err)
+						continue
+					}
+					go func(name string) {
+						defer func() {
+							m.Lock()
+							logging[name] = false
+							m.Unlock()
+							readCloser.Close()
+						}()
+						scanner := bufio.NewScanner(readCloser)
+						for scanner.Scan() {
+							fmt.Fprintf(to, "%s: %s\n", name, scanner.Text())
```

Comment:
Is this going to cause logs from multiple containers to be interspersed with each other?

Reply:
Looking at the output [here](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/69986/pull-kubernetes-e2e-gce/56360/build-log.txt), this seems useful for debugging, but is incredibly verbose and makes it difficult to extrapolate the test steps since all the logs are mixed together.

Is there a way we can dump all these logs to other files and have test-infra include them in the artifacts?  cc @kubernetes/sig-testing-pr-reviews

Reply:
If `TestContext.ReportDir` is nonempty, then it'll point to a directory that's uploaded to GCS after the run. I believe all e2e test runs on Prow set this.

Reply:
Yes, output from multiple containers will be interspersed. But there is a "pod/container" prefix, so one can still determine the origin of the log message.

The log file that @msau42 linked to becomes crowded because Ginkgo seems to run in streaming mode, i.e. all output ends up in the log. In a perfect world, all output would go to the `GinkgoWriter` instead of `os.Stdout`. Then streaming could be turned off and only the output of failed tests would be in the log, with the full set of information.

But we don't live in a perfect world, so we need to find some compromise.

I don't like writing to extra files. Developers who look at a test failure like https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/69868/pull-kubernetes-e2e-gce/56388/#sig-storage-csi-volumes-csi-plugin-test-using-csi-driver-hostpath-should-provision-storage will then have to use the "Filter through log files" feature and manually select the right extra log file(s). The steps in the test (i.e. the `By()` output) won't be there. In theory there would be the advantage that one could use the "Weave by timestamp" feature to mix pod output with the corresponding kubelet output - but in practice I just got a "Server Error" when I tried that for two different `kubelet.log` files in the previous example.

How about this: when the tests run in the CI, they dump the extra information into a memory buffer. If the test fails, that memory buffer gets dumped at the end, similar to the "Collecting events from namespace" feature in framework.go. When invoked by a developer, the extra information is printed normally (streamed or captured by Ginkgo). I've use this code for a while now in my (still closed) project, and it is very useful to see what went wrong as it goes wrong, instead of having to wait for the test to time out and only then print that crucial piece of information.

There will be some implementation challenges, though, so I don't know how well I can make that work:
- How can the code detect the "run in the CI" part and switch modes? Yet another command line parameter? Or can we use the "report dir enabled" as indicator for "run in the CI"?
- The normal output via `GingkoWriter` should be mixed into the memory buffer, otherwise steps are again missing.

Reply:
I actually prefer separating logs into multiple files. If I'm debugging an issue and I suspect the attacher, then I can just pull up the attacher logs and see only those relevant messages without having to mentally filter out the rest of the spam.  You do lose interspersing with the `By()` steps, however, you cannot trust the timing of those messages with the logs anyway since they are all being processed and logged in parallel.

Reply:
Sounds like there are different usage patterns. Let's make writing to separate log files the default if reportdir is enabled, otherwise log to GingkoWriter with a prefix.

What should those logs be called? <test-name>/<pod>/<container>.log, with some mangling to replace spaces and remove special characters?

For example: `sig-storage-CSI-Volumes-CSI-plugin-test-using-CSI-driver-hostPath-should-provision-storage/csi-pod/driver-registrar.log`.

Can the web UI handle multiple nested directories? It seems it can handle <host>/<something>.log.

Reply:
I've implemented this approach, but was not able to test after rebasing because of https://github.com/kubernetes/kubernetes/issues/70185.

Let's see how it works for the gcePD test in the CI...